### ISSUE TITLE
Camera / bluetooth not working on yogamsc branch

### ISSUE THREAD

**Issue by Charlyo** [EETagent/T480-OpenCore-Hackintosh#57]
Hi @EETagent!

Thank you for all your work.

Currently I've gotten a T480 (i5 8350U) and installed your yoga smc branch since I wanted to have YogaSMC.

However, despite having all the necessary USB ports in the XHC.aml file, at least the camera and bluetooth (from a bradcom 4352z) are not working. Both devices work with USBInjectAll. 

Let me know if changes have to be made!

Best regards!

**Comment by EETagent** [comment]
Hi @Charlyo

Thanks for testing experimental branch. I actually did not test webcam and Bluetooth since moving to SSDT-XHC, Bluetooth tray icon on Big Sur is not crossed out when unavailable unlike on Catalina so this change went undetected :) I already edited draft's false description. For now you can remove this patch and add USBMap/USBPorts back.

@benbender 
![image](https://user-images.githubusercontent.com/20557318/99310542-77aea000-285b-11eb-993d-1b0b67d94a7f.png)

HS08, HS07 and SS03 are undetected and neither of them works. How we can help to solve this?

Current HS08 config looks OK to me

```
0xFF - Port is enabled
0xFF - Internal connector type
Zero, 
Zero
```

**Comment by Charlyo** [comment]
Thanks for the update @EETagent!!

Actually, if you thunk it through, only the "internal" ports fail.

**Comment by benbender** [comment]
@EETagent It seems that you copied my config from the X1C6 but it differs from the T480. F.e. S03 is internal and SDCard on the T480 because the T480 only has one TB-port.

**Comment by EETagent** [comment]
Thanks @benbender,

sadly the situation has not changed. These 3 USB devices still does not work.

**Comment by benbender** [comment]
A bit more info would be nice :) IOReg-dump? Logs? Is the SSDT applied? Rename for _UPC to XUPC in place?

**Comment by benbender** [comment]
If your config.plist from the YogaSMC-branch is complete, you are missing the needed patch:

````
			<dict>
				<key>Comment</key>
				<string>XHC: _UPC to XUPC</string>
				<key>Count</key>
				<integer>0</integer>
				<key>Enabled</key>
				<true/>
				<key>Find</key>
				<data>X1VQQw==</data>
				<key>Limit</key>
				<integer>0</integer>
				<key>Mask</key>
				<data></data>
				<key>OemTableId</key>
				<data>UHJvalNzZHQ=</data>
				<key>Replace</key>
				<data>WFVQQw==</data>
				<key>ReplaceMask</key>
				<data></data>
				<key>Skip</key>
				<integer>0</integer>
				<key>TableLength</key>
				<integer>6103</integer>
				<key>TableSignature</key>
				<data>U1NEVA==</data>
			</dict>
````

This patch renames all USB-port-definitions in ACPI and redoes them. Today I refined my XHC-SSDT for the X1C6. It may be worth to adapt those changes.

For the future: please check if patches you adapt from me are cleanly applied. In this case they can't be without that patch as the _UPC-definitions collide. Check with a search for "AE_" in `log show --last boot`.

**Comment by EETagent** [comment]
No _UPC to XUPC rename ...

Will fix that and report back

**Comment by EETagent** [comment]
> For the future: please check if patches you adapt from me are cleanly applied. In this case they can't be without that patch as the _UPC-definitions collide. Check with a search for `AE_` in `log show --last boot`.

it does not seem so

![image](https://user-images.githubusercontent.com/20557318/99387650-a74dbe80-28d4-11eb-96c4-d039cb28dd5e.png)

**Comment by benbender** [comment]
You are patching an SSDT. This is different from patching DSDT as it needs the correct Table-signature and length to be applied because there can (and will) be multiple SSDT's on the system. Both values can be found in the header of the SSDT to be patched. View the original XHC-SSDT with MacIASL and adapt the patch accordingly. As said above, the patch is taken from my X1C6 as I only have that at hand.

Additionally, grep'ing for AE_ only shows that there are errors. To see the message, grep after AppleACPIPlatform in the logs. The actual error-messages are below the lines starting with `AE_`.

PS: I urgently need to do a writeup on how to do basic ACPI-debugging :)

**Comment by EETagent** [comment]
![image](https://user-images.githubusercontent.com/20557318/99388938-9c942900-28d6-11eb-953a-8a6d1a50cd37.png)



> PS: I urgently need to do a writeup on how to do basic ACPI-debugging :)

Yes, that would be great :)

**Comment by benbender** [comment]
As I said - The patch isn't applied because OpenCore can't find the table. 

```
				<key>TableLength</key>
				<integer>6103</integer>
```

Probably only the length needs to be adapted. Please switch the value with the one found in the original SSDT for the T480 in MacIASL.

**Comment by EETagent** [comment]
Length replaced, SSDT still does not run
![image](https://user-images.githubusercontent.com/20557318/99390337-c0586e80-28d8-11eb-9e27-c7caa7463dbe.png)

**Comment by Charlyo** [comment]
@EETagent have you changed OemTableId to the actualy file one?

**Comment by EETagent** [comment]
> @EETagent have you changed OemTableId to the actual file one?

Well, no. Currently there is `ProjSsdt`, will replace it with `_XHC`

**Comment by Charlyo** [comment]
Not sure if necessary, just felt weird (at least to me).

**Comment by EETagent** [comment]
- [x] OemTableId `_XHC`
- [x] TableLenght `1614`
- [x] TableSignature `SSDT`

Rename still does not apply.

**Comment by benbender** [comment]
> Length replaced, SSDT still does not run
> ![image](https://user-images.githubusercontent.com/20557318/99390337-c0586e80-28d8-11eb-9e27-c7caa7463dbe.png)

Thats the length of your SSDT, not the original. The original is ProjSsdt as @Charlyo and should be longer.

**Comment by EETagent** [comment]
> > Length replaced, SSDT still does not run
> > ![image](https://user-images.githubusercontent.com/20557318/99390337-c0586e80-28d8-11eb-9e27-c7caa7463dbe.png)
> 
> Thats the length of your SSDT, not the original. The original is ProjSsdt as @Charlyo said and should be longer.

I am idiot, will fix it immediately

**Comment by EETagent** [comment]
OK, so length of ProjSsdt is 5692

**Comment by EETagent** [comment]
Webcam, Bluetooth and SD Card finally works! Thanks @benbender for all your help

**Comment by EETagent** [comment]
@Charlyo Fixed in 4ef56250749f4ca270d0dd3809bfa6a26f1624b3

Please try and report back

**Comment by Charlyo** [comment]
Working perfectly on last commit!
